truffle/src/main/java/org/jruby/truffle/nodes/core/BigDecimalNodes.java

```diff
@@ -137,36 +199,58 @@ public AddOpNode(RubyContext context, SourceSection sourceSection) {
 
         @CompilerDirectives.TruffleBoundary
         private Object addBigDecimal(RubyBasicObject a, BigDecimal b) {
-            return createRubyBigDecimal(
-                    getContext().getCoreLibrary().getBigDecimalClass(),
-                    getBigDecimalValue(a).add(b));
+            return createRubyBigDecimal(getBigDecimalValue(a).add(b));
         }
 
-        @Specialization
+        @Specialization(guards = "isNormalRubyBigDecimal(a)")
         public Object add(RubyBasicObject a, int b) {
             return addBigDecimal(a, getBigDecimalValue(b));
         }
 
-        @Specialization
+        @Specialization(guards = "isNormalRubyBigDecimal(a)")
         public Object add(RubyBasicObject a, long b) {
             return addBigDecimal(a, getBigDecimalValue(b));
         }
 
-        @Specialization
+        @Specialization(guards = "isNormalRubyBigDecimal(a)")
         public Object add(RubyBasicObject a, double b) {
             return addBigDecimal(a, getBigDecimalValue(b));
         }
 
-        @Specialization
+        @Specialization(guards = "isNormalRubyBigDecimal(a)")
         public Object add(RubyBasicObject a, RubyBignum b) {
             return addBigDecimal(a, getBigDecimalValue(b));
         }
 
-        @Specialization(guards = "isRubyBigDecimal(b)")
-        public Object add(RubyBasicObject a, RubyBasicObject b) {
+        @Specialization(guards = {"isNormalRubyBigDecimal(a)", "isNormalRubyBigDecimal(b)"})
+        public Object addNormal(RubyBasicObject a, RubyBasicObject b) {
             return addBigDecimal(a, getBigDecimalValue(b));
         }
 
+        @Specialization(guards = "isSpecialRubyBigDecimal(a) || isSpecialRubyBigDecimal(b)")
+        public Object addSpecial(RubyBasicObject a, RubyBasicObject b) {
+            final Type aType = getBigDecimalType(a);
+            final Type bType = getBigDecimalType(b);
+
+            if (aType == Type.NAN || bType == Type.NAN ||
+                    (aType == Type.POSITIVE_INFINITY && bType == Type.NEGATIVE_INFINITY) ||
+                    (aType == Type.NEGATIVE_INFINITY && bType == Type.POSITIVE_INFINITY))
+                return createRubyBigDecimal(Type.NAN);
+
+            if (aType == Type.POSITIVE_INFINITY || bType == Type.POSITIVE_INFINITY)
+                return createRubyBigDecimal(Type.POSITIVE_INFINITY);
+
+            if (aType == Type.NEGATIVE_INFINITY || bType == Type.NEGATIVE_INFINITY)
+                return createRubyBigDecimal(Type.NEGATIVE_INFINITY);
+
+            // a,b are either normal or +-zero
```

Comment:
changed to "// a and b have value NEGATIVE_ZERO or NORMAL". POSITIVE_ZERO was removed.